cpp/src/arrow/array-test.cc

```diff
@@ -1154,6 +1154,54 @@ TEST_F(TestBinaryBuilder, TestScalarAppend) {
     }
   }
 }
+ 
+TEST_F(TestBinaryBuilder, TestCapacityReserve) {
+  vector<string> strings = {"a", "bb", "cc", "ddddd", "eeeee"};
+  int64_t N = static_cast<int>(strings.size());
+  int64_t length = 0;
+  int64_t data_length = 0;
+  int64_t capacity = N;
+ 
+  ASSERT_OK(builder_->Reserve(capacity));
+  ASSERT_OK(builder_->ReserveData(capacity));
+  
+  ASSERT_EQ(builder_->length(), length);
+  ASSERT_EQ(builder_->capacity(), BitUtil::NextPower2(capacity));
+  ASSERT_EQ(builder_->value_data_length(), data_length);
+  ASSERT_EQ(builder_->value_data_capacity(), capacity);
+ 
+  for(const string& str : strings) {
+    ASSERT_OK(builder_->Append(str));
+    length++;
+    data_length += static_cast<int>(str.size());
+   
+    ASSERT_EQ(builder_->length(), length);
+    ASSERT_EQ(builder_->capacity(), BitUtil::NextPower2(capacity));
+    ASSERT_EQ(builder_->value_data_length(), data_length);
+    if (data_length <= capacity) {
+      ASSERT_EQ(builder_->value_data_capacity(), capacity);
+    } else {
+      ASSERT_EQ(builder_->value_data_capacity(), data_length);
```

Comment:
It's not clear to me why these assertions would hold true

```
ASSERT_EQ(builder_->capacity(), BitUtil::NextPower2(capacity));
ASSERT_EQ(builder_->value_data_capacity(), capacity);
```

I would think that `value_data_capacity()` is always the power of 2 greater than or equal to the amount of data appended so far, i.e. `ASSERT_EQ(BitUtil::NextPower2(data_length), builder_->value_data_capacity())`

Can you make the strings you are appending much larger, at least 10 length each?

Reply:
@wesm value_data_capacity() is actually always a multiple of 64 greater than or equal to the amount of data appended so far because the underlying buffer size is set to ensure that the capacity of the buffer is a multiple of 64 bytes as defined in Layout.md, i.e.

ASSERT_EQ(BitUtil::RoundUpToMultipleOf64(data_length), builder_->value_data_capacity())

So if you call ReserveData(capacity) at the very beginning, then we have

ASSERT_EQ(BitUtil::RoundUpToMultipleOf64(capacity), builder_->value_data_capacity())